### DISCUSSION TITLE
Getting it to run on Ubuntu running in M1 VMware Fusion

### DISCUSSION THREAD

**Discussion by helje5** [AparokshaUI/Adwaita#5] [Q&A] [answered]
I'm trying to get it to run on an M1 Mac using VMware Fusion.

Here is what I did:
- install Ubuntu Server ARM in VMware, 22.0.4.4
- install Ubuntu Dekstop (`apt install ubuntu-desktop`)
- install adwait-1-dev and Swift dependencies:
  - `apt-get install libadwaita-1-dev binutils git gnupg2 libc6-dev libcurl4-openssl-dev libedit2 libgcc-9-dev libpython3.8 libsqlite3-0 libstdc++-9-dev libxml2-dev libz3-dev pkg-config tzdata unzip zlib1g-dev`
- install Swift 5.10: https://download.swift.org/swift-5.10-release/ubuntu2204-aarch64/swift-5.10-RELEASE/swift-5.10-RELEASE-ubuntu22.04-aarch64.tar.gz
- install Flatpaks `flatpak install flathub $i`:
  -  org.freedesktop.Sdk
  - org.freedesktop.Sdk.Extension.llvm16/aarch64/23.08
  - org.gnome.Sdk (aarch74/3.38)
  - org.freedesktop.Sdk.Extension.swift5
  - org.gnome.Builder
  - org.gnome.design.AppIconPreview
  - org.inkscape.Inkscape
  - com.visualstudio.code

Then I open the template in in Gnome, which downloads another 2GB SDK, and I end up with such:
```
Working copy of https://github.com/apple/swift-syntax resolved at 509.0.0
Creating working copy for https://github.com/david-swift/LevenshteinTransformations
Working copy of https://github.com/david-swift/LevenshteinTransformations resolved at 0.1.2
warning: circular dependency detected while parsing libadwaita-1: harfbuzz -> freetype2 -> harfbuzz
warning: circular dependency detected while parsing libadwaita-1: freetype2 -> harfbuzz -> freetype2
warning: circular dependency detected while parsing libadwaita-1: freetype2 -> harfbuzz -> freetype2
```

```
Building for production...
/home/helge/tmp/AdwaitaTemplate/.build/checkouts/Adwaita/Sources/Adwaita/View/Dialogs/AboutDialog.swift:57:17: error: cannot find 'adw_dialog_present' in scope
                adw_dialog_present(storage.content[dialogID]?.first?.pointer?.cast(), storage.pointer?.cast())
                ^~~~~~~~~~~~~~~~~~
/home/helge/tmp/AdwaitaTemplate/.build/checkouts/Adwaita/Sources/Adwaita/View/Dialogs/AboutDialog.swift:57:79: error: generic parameter 'T' could not be inferred
                adw_dialog_present(storage.content[dialogID]?.first?.pointer?.cast(), storage.pointer?.cast())
                                                                              ^
/home/helge/tmp/AdwaitaTemplate/.build/checkouts/Adwaita/Sources/Adwaita/Model/Extensions/OpaquePointer.swift:12:17: note: in call to function 'cast()'
    public func cast<T>() -> UnsafeMutablePointer<T> {
                ^
/home/helge/tmp/AdwaitaTemplate/.build/checkouts/Adwaita/Sources/Adwaita/View/Dialogs/AboutDialog.swift:57:104: error: generic parameter 'T' could not be inferred
```

Outside of Builder I also get cycle warnings in a unsuspicious looking project, wonder whether the macros do something weird.

**Comment by CmST0us** [comment]
It seems you need to run `Generation` before build

**Reply by david-swift** [reply]
`Generation` does not have to be run before build, as it assumes the newest version of libadwaita is installed as well. It is really recommended to run in a Flatpak where you can use the newest version by simply changing the GNOME version, take a look at the [template repo](https://github.com/AparokshaUI/AdwaitaTemplate).

**Comment by david-swift** [comment, marked as answer]
Dialogs are a new feature in the latest libadwaita release. I didn't update the template (will do), so you need to change [this line](https://github.com/AparokshaUI/AdwaitaTemplate/blob/7fc6f7b5fc69fd044138c6fb6da2a711651efef9/io.github.AparokshaUI.AdwaitaTemplate.json#L4) to use the runtime version 46. 

Also, you don't have to install any dependencies (including Swift) apart from Flatpaks on your system if you are building the app in a Flatpak environment (which is recommended). I'll update the template repo and also this repo to ensure that people don't misunderstand and think they have to install all the dependencies when using Flatpaks.

**Reply by david-swift** [reply]
I've updated the Flatpak manifest to GNOME version 46 as well as the [instructions](https://github.com/AparokshaUI/AdwaitaTemplate?tab=readme-ov-file#installation) to make clear that dependencies do not have to be installed on the system.

**Reply by david-swift** [reply]
The installation section in this repo's README now contains information about Flatpak as well.

**Reply by helje5** [reply]
OK, that worked! Thanks a lot.

**Reply by unnamedd** [reply]
@david-swift would be awesome if some instruction on Asahi Linux (Fedora) would be given too. I am trying to make it work, but with no success. Builder for some reason that I cannot understand well, cannot install the dependencies.

**Reply by helje5** [reply]
I think the short version is that it isn't distri dependent at all, it all runs in `flatpak`. You follow the instructions and it should work. Make sure to stay in flatpak. (it would be nice to get plain-distro instructions too, but that is probably not viable)

**Reply by david-swift** [reply]
@unnamedd Make sure that the active configuration is on Flatpak and not on Swift. Does the app's id get displayed in the "Build Menu" (next to the hammer) > Active Configuration? Otherwise, try running it with `flatpak-builder run io.github.AparokshaUI.AdwaitaTemplate.json`. Install the dependencies manually `flatpak install org.freedesktop.Sdk.Extension.swift5`, `flatpak install org.gnome.Platform`. If you aren't able to install the dependencies, set up [Flathub](https://flathub.org/setup) first.

**Reply by unnamedd** [reply]
@david-swift thanks a lot for the instructions here, I had no idea on how to try to run them directly on terminal. Will return to you later!

**Reply by unnamedd** [reply]
@helje5 Sorry, I didn't notice your message right before the one from David. The biggest problem for me was: I didn't know what kind of behavior expect from Builder. When running `swift build`, I was killing the process thinking it was not working, but in the first time it takes more time to compile and run than I initially expected. 

Being on a Linux running on Apple Silicon some times is a bit tricky and things doesn't work as expected and that was my initial thought. By a fortunate accident, on another Linux (Intel)  it worked nicely because I forgot the compiling was running.  When I tried again on Asahi and it worked.

Lesson learned? Absolutely! 
If you happen to come across this comment in the future, don't be discouraged by the initial long run time; it will eventually work. If not, Swift will provide feedback.

**Reply by helje5** [reply]
Hail Swift Macros, it is libSyntax which your are building for hours ;-)

**Reply by david-swift** [reply]
Swift macros are used for an alternative deprecated approach to localization using the [Localized package](https://github.com/AparokshaUI/Localized). I removed this approach in the latest release so it should be faster now.
